### DISCUSSION TITLE
SQL for document similarity in edge function

### DISCUSSION THREAD

**Discussion by apurbanand** [supabase/supabase#21051] [Questions] [answered]
Why is `documentsWithSimilarity` being returned as a blank array in the code below irrespective of any value of  `matchThreshold`?

        
 ```
        const result = await connection.queryObject({
          text: `SELECT
            documents_embeddings_table.document_content,
            1.00 - (documents_embeddings_table.embeddings <=> $.7) as similarity
          FROM documents_embeddings_table
          WHERE documents_embeddings_table.embeddings <=> $1 < 1.00 - CAST($2 AS float)
          LIMIT $3`,
          args: ['[' + questionEmbedding + ']', matchThreshold, matchCnt]
        });

        console.log('Query:', result.query);
        console.log('Values:', result.values);
        console.log('Result:', result.rows);

        const documentsWithSimilarity = result.rows;
        console.log('documentsWithSimilarity.length');
        console.log(documentsWithSimilarity.length);
        console.log('documentsWithSimilarity');
        console.log(documentsWithSimilarity);

        // Log individual similarity values
        console.log('Individual Similarity Values:');
        documentsWithSimilarity.forEach((doc, index) => {
          console.log(`Document ${index + 1}: ${doc.similarity}`);
        });

```
Any help on this will greatly appreciated please.

**Comment by mwoss** [comment]
Have you correctly set up RLS for the `documents_embeddings_table` table? Incorrect RLS policies are usually the main reason for problems like you have.

**Reply by apurbanand** [reply]
Thanks for early response @mwoss . The  

> documents_embeddings_table

  is public schema with no RLS policy as yet.

**Reply by mwoss** [reply]
If you want to allow public access to this table you would need to create an appropriate RLS policy then :3 Take a look here - [Supabase RLS documentation](https://supabase.com/docs/guides/auth/row-level-security#always-enable-rls-on-public-tables). Putting table in the public schema is not enough.

**Reply by apurbanand** [reply]
Ok.
Will do that and get back.
Thanks!!

**Reply by apurbanand** [reply]
I have enabled RLS policy on the  `documents_embeddings_table` but still getting blank array for `documentsWithSimilarity`. I am reproducing the logs on the Edge Functions Logs page-

```
Query: Query {
  args: [
    "[0.0010733653,-0.029097276,0.020605681,-0.02195495,-0.029874573,-0.03161982,-0.033203747,0.019975046"... 19138 more characters,
    "0.6",
    "5"
  ],
  camelcase: undefined,
  fields: undefined,
  result_type: 1,
  text: "SELECT\n" +
    "            documents_embeddings_table.document_content,\n" +
    "            1.00 - (documents_embedd"... 192 more characters
}
------------------
Values: undefined
------------------
Result: []
---------------

documentsWithSimilarity.length = 0
------------------
documentsWithSimilarity = []
```

**Reply by mwoss** [reply]
If you run the same query in the Supabase SQL editor do you receive some values? Could you share the RLS policy you've applied on this table too? :D

**Reply by apurbanand** [reply]
`SELECT documents_embeddings_table.document_content, (1.00 - (documents_embeddings_table.embeddings <=> $1::vector)) FROM documents_embeddings_table WHERE documents_embeddings_table.embeddings <=> $1::vector < 1.00 - CAST($2 AS float) LIMIT $3`

Since  `documents_embeddings_table.embeddings` is 1536 dimensional vector, I am not sure how to run the query in Supabase SQL editor. The RLS policy that I applied on table is 
```
create policy "Allow public access"
on documents_embeddings_table for select
using ( true );
```

**Reply by mwoss** [reply]
RLS looks good in my opinion. Could you just run `SELECT * FROM documents_embeddings_table` to verify it's not an issue with data access? :3

**Reply by apurbanand** [reply, marked as answer]
After a lot of experiments, I ultimately got it running with....

```
        // Run a query with custom logic
        const result = await connection.queryObject({
          text: `SELECT documents_embeddings_table.document_content, documents_embeddings_table.embeddings <=> $1::vector as similarity FROM documents_embeddings_table WHERE documents_embeddings_table.embeddings <=> $1::vector > CAST($2 AS float) ORDER BY documents_embeddings_table.embeddings <=> $1::vector LIMIT $3`,
          args: [JSON.stringify(questionEmbedding), matchThreshold, matchCnt]
        });
```

Thanks for your support @mwoss.
